### DISCUSSION TITLE
Redacting text while preserving layout

### DISCUSSION THREAD

**Discussion by djoltes** [pymupdf/PyMuPDF#2195] [Looking for help]
Need some ideas; I'm relatively new to PuyMuPDF but have a requirement to perform redaction on files that may or may not contain image pages. Redacting text pages is trivial and I have it working well, but some files contain images of invoices or other documents and I need to preserve the existing layout so I can't just grab the text layer.

Has anyone developed a process that will detect 'image' pages, then route that page into a handler to OCR the page and return the full output so redaction can be performed against the extracted text? I looked at the 'ocrpages.py' sample that uses ocrmypdf, but keep getting errors during extraction -- I have both Tesseract and GSview installed (Windows 2019 server) and they've been added to the PATH, and can't figure out why I get the 'cannot find the file specified' errors.

Text Length: 1

Scanning contents:   0%|          | 0/1 [00:00<?, ?page/s]
Scanning contents: 100%|##########| 1/1 [00:00<00:00, 51.13page/s]

OCR:   0%|          | 0.0/1.0 [00:00<?, ?page/s][tesseract] lots of diacritics - possibly poor OCR

OCR:  50%|#####     | 0.5/1.0 [00:05<00:05, 10.43s/page]
OCR: 100%|##########| 1.0/1.0 [00:05<00:00,  5.22s/page]

Recompressing JPEGs: 0image [00:00, ?image/s]
Recompressing JPEGs: 0image [00:00, ?image/s]

Deflating JPEGs:   0%|          | 0/1 [00:00<?, ?image/s]
Deflating JPEGs: 100%|##########| 1/1 [00:00<00:00, 50.02image/s]

[WinError 2] The system cannot find the file specified
[WinError 2] The system cannot find the file specified

Any leads appreciated...

**Comment by JorjMcKie** [comment]
There are several indicators suggesting that a page "is" an image and has no text:

1. There exists (at least one) an image covering the complete page rectangle.
2. `page.get_text()` delivers an empty string.
3. If font `GlyphLessFont` is being used on your page, then you know that the page was OCRed using Tesseract

Check 1. like this:
```python
imgxrefs=[item[0] for item in page.get_images()]
rects = []  # list of all rectangles on page covered by some image
for xref in imgxrefs:
    rects.extend(page.get_image_rects(xref))
for r in rects:
    if page.rect in r:
        print("page covered by an image")
        break
# to check if a page is at least "roughly" covered, check this:
    if abs(page.rect - r) < some_threshold:   # like some_threshold = 0.1 or so
        print("page covered by an image")
        break
```

Check 3. like this:
```python
ocred = [item[3] for item in page.get_fonts() if "GlyphLessFont" in item[3]] != []
```
If false, then the page was not OCRed (with Tesseract at least).

But please consider this:
If you redact an OCRed page, then the text will be redacted for sure, but the image also needs to be modified to remove the text part inside the image. For this the image will be extracted and white / empty rectangles will be put in it at all placed where there is redacted text.
This procedure takes a relatively long time, so make sure to do this only once per page: add redaction annotations for all text pieces to remove, **then** execute `page.apply_redactions()`.

**Reply by djoltes** [reply]
Interesting, and thanks for the reply. I've actually made progress in the respect of being able to pass an 'image' page to a method to perform the OCR, get the returned result, then perform redactions against it. As a starting point I have the following code:

[note: ocr_the_page() is basically the same as what's in the ocrpages.py file, except that instead of returning the OCR'd text it's returning the in-memory PDF, e.g., 

   ```
 ocr_pdf = fitz.open("pdf", outbytes.getvalue())  # read output as fitz PDF
    return ocr_pdf

 for page in doc:
        page.wrap_contents()
        print("Page #: {a}".format(a = pageNo))
        pageContent = page.get_text("text").split("\n")

        if len(pageContent) <= 1:
            # This means we need to OCR the page first...
            try:
                pageExtract = ocr_the_page(page)
                page = pageExtract[0]  **<-- is this the correct way to handle replacing the existing page with the OCR'd version?**
                pageContent = page.get_text("text").split("\n")
```

I can see the redactions being performed against pageContent afterward, and the annotations are applied in the same way as with a normal 'text' page.  But once they're applied and the PDF is written, the 'image' pages are still images and no redaction has been done. What am I doing wrong here?

**Reply by JorjMcKie** [reply]
> But once they're applied and the PDF is written, the 'image' pages are still images and no redaction has been done. 

Well, applying redactions modifies images where they are intersecting redaction rectangles - this is the default option.
You **_can__** also completely remove every intersecting image of completely leave it untouched.
If you use the default and intersecting images were not modified, this would be a bug.

**Reply by djoltes** [reply]
So is my method of retrieving an OCR'd page object and replacing the existing page object in the loop with it correct? Again, I know that page object now contains the OCR'd text because I can see it being examined by my redaction code (the "Checking..." log entries):

Checking: Avutomonve Ap STRATION
Checking: Leonardo Marchan

Part of the redaction code is as follows. Again, this works perfectly for text pages but when I replace an 'image' page with its OCR'd equivalent the resulting PDF just contains the original image pages. 

  ```
ents = getEnts(pageContent)
        if ents:
            for ent in ents:
                e1 = str(ent)
                if any(ele in e1 for ele in exclude_list):
                    print("Avoiding redaction of: {a}".format(a = e1))
                    continue
                print("Checking: {a}".format(a=e1))
                areas = page.search_for(e1)         
                # draw outline over sensitive data
                [page.add_redact_annot(area, fill = (0, 0, 0)) for area in areas]
        
        # apply the redactions
        page.apply_redactions()
```

**Reply by JorjMcKie** [reply]
I cannot debug your code, but I am interested in whether images are correctly showing "holes" where redaction rectangles existed.
The following shows a successful removal of text "Documentation" from an image-only, OCRed page:
Before: 
![grafik](https://user-images.githubusercontent.com/8290722/215151462-6bff8c6d-bdcc-4351-964c-d50941bceb76.png)
After: 
![grafik](https://user-images.githubusercontent.com/8290722/215153226-90980601-66df-4c9f-b94c-08b54c627bf6.png)

**Reply by djoltes** [reply]
I'm not asking for anyone to debug it, I just need to know what process will work on a multi-page PDF that contains a mix of image and text pages. What I'm doing now, as noted above, is sending the image pages through OCR and returning the reprocessed page, which clearly does contain OCR text I can redact. The only remaining problem is replacing the image page with the OCR page, which doesn't seem to work using my existing method. 

If someone has a solution to this, I'd very much appreciate it.

**Reply by JorjMcKie** [reply]
Don't get me wrong: it seems we haven't exchanged all information, or I didn't understand you yet. Let me restate what I understood so far:
1. You have identified a page consisting of an image only
2. you then OCRed it and successfully (!) remove certain text, including removal from the image
3. and now what is the problem: The page has achieved the state you wanted?

**Reply by djoltes** [reply]
Okay, let me restate where I am and how I'm going about this.

1) I iterate over pages as usual. If a page is already text, it's just processed normally. This is working fine and redactions appear.
2) if a page is detected as image, it's sent to the OCR method (ocr_the_page()) and the resulting PDF object is returned.
```
ocr_pdf = fitz.open("pdf", outbytes.getvalue())  # read output as fitz PDF
return ocr_pdf
```
3) the 'page' variable in the main logic's loop ("for page in doc:")  is replaced with the first page from the returned OCR'd object 
```
pageExtract = ocr_the_page(page)
page = pageExtract[0]
```
4) I now extract the text from the new page object, which works so it means that object contains OCR'd text
5) annotations are performed against the extracted text
6) page.apply_redactions() is performed to ostensibly apply the changes

However, when I view the resulting PDF the image pages are still images and no redaction has occurred on them. So it appears that simply replacing the 'page' variable with the OCR'd version isn't working for some reason. This is where I need ideas...maybe I'm doing this incorrectly?

**Reply by JorjMcKie** [reply]
Ah, I believe I have it now.
I think you have two options:
1. When you have created that PDF with the OCRed pixmap, nothing prevents you from redacting its single page right away there. Then copy it over to your original PDF. For this copying, you must use `target_pdf.insert_pdf(ocr_pdf, start_at=page.number+1)`. This will make the OCRed page immediately follow tha current page. Then do `target_pdf.delete_page(page.number)` to remove the current page - this lets the OCRed page take over its place.
2. If do not really need the image page to contain OCRed text, but you simply want to clean the text from the image, you can still make an redaction annotation for it and let the image be modified. You can use the same rectangle that you computed for the ocr_pdf.

As you have observed: re-assigning the Python `page` variable to something else has no influence.

**Reply by djoltes** [reply]
Ah, that's what I was looking for...and yes, it makes sense that you can't just replace the object. That said, I get an error from the library on the delete attempt. 

 ```
       if len(pageContent) <= 1:
            # This means we need to OCR the page first...
            try:
                pageExtract = ocr_the_page(page)
                doc.insert_pdf(pageExtract, start_at=page.number+1)
                doc.delete_page(page.number)
                continue
```

Traceback (most recent call last):
  File "C:\Analytics\workspace\WeeklyInjury\src\PDFredact.py", line 158, in <module>
    doc.delete_page(page.number)
  File "C:\Program Files\python_310\lib\site-packages\fitz\fitz.py", line 5281, in delete_page
    while pno < 0:
TypeError: '<' not supported between instances of 'NoneType' and 'int'

**Reply by JorjMcKie** [reply]
> `doc.delete_page(page.number)`

Ah! Because inserting pages from whereever, all existing page objects are being invalidated - as a precaution against invalidation of page numbers. I tricked myself out here 😉.
So you must save the number of the current page into some variable before `insert_pdf()` which you then use here.

**Reply by djoltes** [reply]
LOL, I hate when that happens! Okay, this seems to be working -- at least I'm now getting the OCR'd texts in the output PDF, so thanks very much for the chat. 

This is all part of a project to redact texts automatically as much as possible -- currently I have handlers for phone numbers, vehicle IDs (VINs), email addresses, and I'm using NER tools from spaCy to detect proper names. We'll see what the accuracy level turns out to be...

**Reply by JorjMcKie** [reply]
> LOL, I hate when that happens!

😎 How much more would you hate to see your Python interpreter crashing because of referencing a page that no longer exists ...
